### PR TITLE
Turn screen off when closing scrcpy

### LINKED ISSUE #714: Power off screen on close
Would it be possible to have a flag power off the screen when you close the stream, like how you do with Ctrl-P, but it activates when you close the scrcpy window.

### PR DESCRIPTION
**Pull request by npes87184**
Fix #714

### CODE REVIEW

File: app/src/control_msg.h [hunk 1]
```diff
@@ -25,6 +25,7 @@ enum control_msg_type {
     CONTROL_MSG_TYPE_GET_CLIPBOARD,
     CONTROL_MSG_TYPE_SET_CLIPBOARD,
     CONTROL_MSG_TYPE_SET_SCREEN_POWER_MODE,
+    CONTROL_MSG_TYPE_TURN_SCREEN_OFF,
```
Comment (rom1v) [inline app/src/control_msg.h]: We need to distinguish between "turn screen off but keep device on" (like [here](https://github.com/Genymobile/scrcpy/blob/513d1ac96d24d6cf07fb1850232b5c0e8f555615/app/src/scrcpy.c#L391)) and "turn screen off and device off (standby)". (There is already some naming confusion in the current codebase, if you have any suggestion…)

**Comment by rom1v** [comment]
I merged the first commit fixing the log :+1:

File: app/src/control_msg.h (hunk 1 above)
Comment (npes87184) [reply app/src/control_msg.h]: Maybe lock screen vs turn screen off?

**Comment by npes87184** [comment]
I have pushed v2:
* Let this feature be optional.
* use "lock screen"

However, I found that server may be killed before screen off.
We may need to introduce some ack mechanism from server to client.

**Comment by rom1v** [comment]
> However, I found that server may be killed before screen off.
> We may need to introduce some ack mechanism from server to client.

In my TODO list, I plan to implement something similar to #679 to be able to cleanup properly:
> The server "forks" itself for full life cycle control (adb may kill it instead preventing required cleanup).

This will be a requirement to merge this PR.

**Comment by rom1v** [comment]
Now that there is a real cleanup, you could implement it :wink:

(I suggest `--power-off-on-close`)

**Comment by rom1v** [comment]
@npes87184 I think several people are interested:
 - #714
 - #1764
 - #2103
 - https://www.reddit.com/r/scrcpy/comments/ljzpdd/turning_off_the_display_automatically_upon/
:wink:

**Comment by npes87184** [comment]
Ok, I will look it.

**Comment by npes87184** [comment]
Hi, @rom1v 

There is a problem about power off screen in `CleanUp` process. In this time, `injectEvent` will consider about `displayId` to inject to correct target. However, it is diffiuclt to allocate `Device` in the process of `CleanUp`. As an result, I cannot trigger power off screen in that stage. Do you have any good advice?

Thank you.

**Comment by rom1v** [comment]
As a quick-and-dirty solution, maybe you could expose a static `injectEvent()` taking a displayId:

```java
diff --git a/server/src/main/java/com/genymobile/scrcpy/Device.java b/server/src/main/java/com/genymobile/scrcpy/Device.java
index a8fdf677..9fb365ff 100644
--- a/server/src/main/java/com/genymobile/scrcpy/Device.java
+++ b/server/src/main/java/com/genymobile/scrcpy/Device.java
@@ -157,16 +157,20 @@ public final class Device {
         return supportsInputEvents;
     }
 
+    public static boolean injectEvent(InputEvent inputEvent, int mode, int displayId) {
+        if (displayId != 0 && !InputManager.setDisplayId(inputEvent, displayId)) {
+            return false;
+        }
+
+        return SERVICE_MANAGER.getInputManager().injectInputEvent(inputEvent, mode);
+    }
+
     public boolean injectEvent(InputEvent inputEvent, int mode) {
         if (!supportsInputEvents()) {
             throw new AssertionError("Could not inject input event if !supportsInputEvents()");
         }
 
-        if (displayId != 0 && !InputManager.setDisplayId(inputEvent, displayId)) {
-            return false;
-        }
-
-        return SERVICE_MANAGER.getInputManager().injectInputEvent(inputEvent, mode);
+        return injectEvent(inputEvent, mode, displayId);
     }
 
     public boolean injectEvent(InputEvent event) {
```

Then, you have to transmit both the displayId and the flag to NOT turn off if control is disabled.

What do you think?

**Comment by npes87184** [comment]
> As a quick-and-dirty solution, maybe you could expose a static `injectEvent()` taking a displayId:
> 
> ```java
> diff --git a/server/src/main/java/com/genymobile/scrcpy/Device.java b/server/src/main/java/com/genymobile/scrcpy/Device.java
> index a8fdf677..9fb365ff 100644
> --- a/server/src/main/java/com/genymobile/scrcpy/Device.java
> +++ b/server/src/main/java/com/genymobile/scrcpy/Device.java
> @@ -157,16 +157,20 @@ public final class Device {
>          return supportsInputEvents;
>      }
>  
> +    public static boolean injectEvent(InputEvent inputEvent, int mode, int displayId) {
> +        if (displayId != 0 && !InputManager.setDisplayId(inputEvent, displayId)) {
> +            return false;
> +        }
> +
> +        return SERVICE_MANAGER.getInputManager().injectInputEvent(inputEvent, mode);
> +    }
> +
>      public boolean injectEvent(InputEvent inputEvent, int mode) {
>          if (!supportsInputEvents()) {
>              throw new AssertionError("Could not inject input event if !supportsInputEvents()");
>          }
>  
> -        if (displayId != 0 && !InputManager.setDisplayId(inputEvent, displayId)) {
> -            return false;
> -        }
> -
> -        return SERVICE_MANAGER.getInputManager().injectInputEvent(inputEvent, mode);
> +        return injectEvent(inputEvent, mode, displayId);
>      }
>  
>      public boolean injectEvent(InputEvent event) {
> ```
> 
> Then, you have to transmit both the displayId and the flag to NOT turn off if control is disabled.
> 
> What do you think?

Oh, yes. This solution works. However, I am thinking about future. It may need to more and more things that need to be done in `CleanUp`. If we need to pass variable likes this in each time, it is very annoy and dirty.

For infrastructure, in my opinion, in `CleanUp` process, I prefer to prepare a environment similar to the main process.
It just needs to handle the `options` correctly. Maybe it can add two methods, `toFile` and `fromFile`, to the class options.

What do you think?

**Comment by rom1v** [comment]
> It just needs to handle the `options` correctly.

It could be great to serialize some structure between the main process and the cleanup process indeed.

However, this structure should not be the existing `Options`, which is neither necessary (not all fields are meaningful for the cleanup process) nor sufficient (the cleanup actions may for example depend on the initial state of the device, which is not stored in the options).

(I don't know if in the end such a structure would be helpful compared to passing each option per argument.)

But yes, it would be great to share the common part of the `Device` initialization between the main and cleanup process. For now, `Device` is quite a mess, and it's not clear semantically what this class represents.

**Comment by npes87184** [comment]
Ok, I will pass `displayId` in this time. The refactoring of `Device` and the environment preparation of `CleanUp` should be another problems (or features).

Thank you for your help.

**Comment by npes87184** [comment]
Hi, @rom1v 

I have added two commits about this features.

First, I export a series static methods for powering off. Second, make `CleanUp` process respects the power off setting.

Thanks.

**Comment by rom1v** [comment]
Thank you very much for your work :+1:

I just tested, and I found a problem (on a Nexus 5): the "power off" cleanup interacts badly with the "restore normal power mode" cleanup, which is executed whenever the screen is on on exit. As a result, `--power-off-on-close` works randomly (on this device).

I tried with these changes:

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/CleanUp.java b/server/src/main/java/com/genymobile/scrcpy/CleanUp.java
index 7dda83fb..26012409 100644
--- a/server/src/main/java/com/genymobile/scrcpy/CleanUp.java
+++ b/server/src/main/java/com/genymobile/scrcpy/CleanUp.java
@@ -84,16 +84,14 @@ public final class CleanUp {
             }
         }
 
-        if (restoreNormalPowerMode) {
-            Ln.i("Restoring normal power mode");
-            if (Device.isScreenOn()) {
+        if (Device.isScreenOn()) {
+            if (powerOffScreen) {
+                Ln.i("Power off screen");
+                Device.powerOffScreen(displayId);
+            } else if (restoreNormalPowerMode) {
+                Ln.i("Restoring normal power mode");
                 Device.setScreenPowerMode(Device.POWER_MODE_NORMAL);
             }
         }
-
-        if (powerOffScreen) {
-            Ln.i("Power off screen");
-            Device.powerOffScreen(displayId);
-        }
     }
 }
```

It improves the behavior.

But I have an additional issue that can I cannot explain: if I keep touching the device (I keep one finger on the device screen) while I close the scrcpy window, it blocks and never power off the screen. It is unexpected, because if I keep one finger on the screen and press <kbd>MOD</kbd>+<kbd>p</kbd> to power off the screen, then it works. :confused: 

If I add some logs (and even if I always return `true`) in `Device.isScreenOn()`:

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/CleanUp.java b/server/src/main/java/com/genymobile/scrcpy/CleanUp.java
index 26012409..2c54558c 100644
--- a/server/src/main/java/com/genymobile/scrcpy/CleanUp.java
+++ b/server/src/main/java/com/genymobile/scrcpy/CleanUp.java
@@ -84,7 +84,9 @@ public final class CleanUp {
             }
         }
 
+        Ln.i("CLEANUP #1");
         if (Device.isScreenOn()) {
+            Ln.i("CLEANUP #2");
             if (powerOffScreen) {
                 Ln.i("Power off screen");
                 Device.powerOffScreen(displayId);
@@ -93,5 +95,6 @@ public final class CleanUp {
                 Device.setScreenPowerMode(Device.POWER_MODE_NORMAL);
             }
         }
+        Ln.i("CLEANUP #3");
     }
 }
diff --git a/server/src/main/java/com/genymobile/scrcpy/Device.java b/server/src/main/java/com/genymobile/scrcpy/Device.java
index 624c9fa0..3ee248d3 100644
--- a/server/src/main/java/com/genymobile/scrcpy/Device.java
+++ b/server/src/main/java/com/genymobile/scrcpy/Device.java
@@ -212,7 +212,7 @@ public final class Device {
     }
 
     public static boolean isScreenOn() {
-        return SERVICE_MANAGER.getPowerManager().isScreenOn();
+        return true; //SERVICE_MANAGER.getPowerManager().isScreenOn();
     }
 
     public synchronized void setRotationListener(RotationListener rotationListener) {
```

Then it blocks:

```
02-21 11:24:23.251 21036 21056 D scrcpy  : Device message sender stopped
02-21 11:24:23.485 21054 21054 I scrcpy  : Cleaning up
02-21 11:24:23.485 21054 21054 I scrcpy  : CLEANUP #1
```

If I replace `Device.isScreenOn()` by `true`, then it prints `CLEANUP #2`, but and `Power off screen`, but does not power off the screen. :thinking:

**Comment by rom1v** [comment]
In fact, it seems the cleanup process is killed (the process does not appear in `ps`). 

Btw, if I add a `SystemClock.sleep(1000);` (even if I don't touch the device), the process is closed before `CLEANUP #1`.
The fact that I touch the screen is irrelevant (it probably just change the timings and makes it more probable).

It's probably due to the `OMXNodeInstance: !!! Observer died`:
```
02-21 11:45:31.420 22574 22593 D scrcpy  : Controller stopped
02-21 11:45:32.469 22592 22592 I scrcpy  : Cleaning up
02-21 11:45:32.478 28310 29411 E OMXNodeInstance: !!! Observer died. Quickly, do something, ... anything...
02-21 11:45:32.590 28310 29411 W GraphicBufferSource: Dropped back down to Loaded without Executing
02-21 11:45:32.624 28310 29411 I OMX-VENC: Component Deinit
```

**Comment by npes87184** [comment]
> In fact, it seems the cleanup process is killed (the process does not appear in `ps`).
> 
> Btw, if I add a `SystemClock.sleep(1000);` (even if I don't touch the device), the process is closed before `CLEANUP #1`.
> The fact that I touch the screen is irrelevant (it probably just change the timings and makes it more probable).
> 
> It's probably due to the `OMXNodeInstance: !!! Observer died`:
> 
> ```
> 02-21 11:45:31.420 22574 22593 D scrcpy  : Controller stopped
> 02-21 11:45:32.469 22592 22592 I scrcpy  : Cleaning up
> 02-21 11:45:32.478 28310 29411 E OMXNodeInstance: !!! Observer died. Quickly, do something, ... anything...
> 02-21 11:45:32.590 28310 29411 W GraphicBufferSource: Dropped back down to Loaded without Executing
> 02-21 11:45:32.624 28310 29411 I OMX-VENC: Component Deinit
> ```
> 
> (I don't have this problem on my OnePlus 7 Pro)

Hmm, I am not familiar with `OMXNodeInstance: !!! Observer died`.

BTW, I don't notice this problem in my note10+ with android 11.


```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/CleanUp.java b/server/src/main/java/com/genymobile/scrcpy/CleanUp.java
index 7dda83fb..26012409 100644
--- a/server/src/main/java/com/genymobile/scrcpy/CleanUp.java
+++ b/server/src/main/java/com/genymobile/scrcpy/CleanUp.java
@@ -84,16 +84,14 @@ public final class CleanUp {
             }
         }
 
-        if (restoreNormalPowerMode) {
-            Ln.i("Restoring normal power mode");
-            if (Device.isScreenOn()) {
+        if (Device.isScreenOn()) {
+            if (powerOffScreen) {
+                Ln.i("Power off screen");
+                Device.powerOffScreen(displayId);
+            } else if (restoreNormalPowerMode) {
+                Ln.i("Restoring normal power mode");
                 Device.setScreenPowerMode(Device.POWER_MODE_NORMAL);
             }
         }
-
-        if (powerOffScreen) {
-            Ln.i("Power off screen");
-            Device.powerOffScreen(displayId);
-        }
     }
 }
```

I will apply this change as it is more correct.

Thank you for your test and survey.

**Comment by rom1v** [comment]
Thank you  :+1: Merged into `dev`

**Pull request closed without merging**